### ISSUE TITLE
How to add samples in batches?

### ISSUE THREAD

**Issue by xiankgx** [nmslib/nmslib#500]
I have some 10 M vectors and I tried to build an index by batch. Here is what I have and they are wrong.

```
# Instantiate index
index = nmslib.init(method="hsnw", space="cosinesimil")

**Several calls** to: index.addDataPointBatch(embeddings)

# Create index
index.CreateIndex().
```

Using only a single batch at a time works. I found out this is a wrong way of doing things as I got wrong results even when using the "brute_force" method.

What is the proper way to do this? Concat the samples to a huge one and do it all in a single batch?

**Comment by searchivarius** [comment]
@xiankgx do you have a reproducible examples? I just checked that using multiple `addDataPointBatch` is designed to expand the index properly (i.e., each batch just adds more data).

**Comment by xiankgx** [comment]
@searchivarius Yes, each call to addDataPointBatch() does add more data to the index, the length of the index is correct as shown in the console log after call to CreateIndex(). However, the returned results during query is wrong even with the brute_force method which should give me exact search. Just concatenating everything to one huge matrix with brute force method gives expected result during query. Regarding sharing of data, how do I share such huge data?

**Comment by xiankgx** [comment]
I also build from source to better utilize the hardware if that means anything.

**Comment by searchivarius** [comment]
Is this issue reproducible with 100K or 1M entries? Is there a chance you have a lot of duplicates?

**Comment by xiankgx** [comment]
These embeddings are image features, created from the same number of files. I have not checked the images for exact duplicates, but there should be similar images. Let me see if I can make you a small reproducible example.

**Comment by xiankgx** [comment]
Here is a link to some (4) pickle files containing image embeddings and image ids/list. And the following is a sample Python script to reproduce the error.

`import argparse
import glob
import pickle

import nmslib
import numpy as np
from tqdm import tqdm


def create_index_by_batch(files, method, space):
    """
    Args
        files: Pickle files containing image embeddings and image list.
        method: nmslib method
        space: nmslib space 
    """

    print(
        f"Instantiating nmslib index of method: {method} and space: {space}")
    index = nmslib.init(method=method, space=space)
    all_images = []

    for p in tqdm(files):
        data = pickle.load(open(p, "rb"))
        images = data["images"]
        embeddings = data["embeddings"]
        assert len(images) == len(embeddings)
        tqdm.write(f"embeddings shape: {embeddings.shape}")

        index.addDataPointBatch(embeddings)
        all_images += images

        del embeddings, images, data

    # Create index
    index.createIndex(print_progress=True)

    return index, all_images


def create_index_whole(files, method, space):
    """
    Args
        files: Pickle files containing image embeddings and image list.
        method: nmslib method
        space: nmslib space 
    """

    print(
        f"Instantiating nmslib index of method: {method} and space: {space}")
    index = nmslib.init(method=method, space=space)
    all_images = []
    all_embeddings = []

    for p in tqdm(files):
        data = pickle.load(open(p, "rb"))
        images = data["images"]
        embeddings = data["embeddings"]
        assert len(images) == len(embeddings)
        tqdm.write(f"embeddings shape: {embeddings.shape}")

        all_images += images
        all_embeddings.append(embeddings)

        del embeddings, images, data

    all_embeddings = np.concatenate(all_embeddings, axis=0)
    index.addDataPointBatch(all_embeddings)

    # Create index
    index.createIndex(print_progress=True)

    return index, all_images


def main(args):
    files = sorted(glob.glob(args.path))
    print(f"embedding files: {len(files)}")

    # Load first embeddings file and get first 5 vectors as queries
    data = pickle.load(open(files[0], "rb"))
    embeddings = data["embeddings"]
    # print(f"embeddings.shape: {embeddings.shape}")
    query = embeddings[:5]
    print(f"query.shape: {query.shape}")

    index1, images1 = create_index_by_batch(files,
                                            method=args.method,
                                            space=args.space)
    index2, images2 = create_index_whole(files,
                                         method=args.method,
                                         space=args.space)

    k = 25
    res1 = index1.knnQueryBatch(query, k=k)
    res2 = index2.knnQueryBatch(query, k=k)

    ids1 = np.array([t[0] for t in res1])
    ids2 = np.array([t[0] for t in res2])
    # print(ids1)
    # print(ids2)

    assert images1 == images2, "Image list not equal."
    assert np.all(ids1 == ids2), "Nearest neighbors not equal."


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--path",
                        type=str,
                        default="./embeddings-part*.pkl")
    parser.add_argument("--method", type=str, default="brute_force")
    parser.add_argument("--space", type=str, default="cosinesimil")
    args = parser.parse_args()

    main(args)`

**Comment by searchivarius** [comment]
@xiankgx thanks a lot for the bug report with reproducible data. Regretfully, there is a bug in how IDs are assigned. It was spotted only now :-( Needs to be fixed ASAP, as a workaround, one has to explicitly specify the list of IDs as in this modification of your code:
```
def create_index_by_batch(files, method, space):
    """
    Args
        files: Pickle files containing image embeddings and image list.
        method: nmslib method
        space: nmslib space 
    """

    print(
        f"Instantiating nmslib index of method: {method} and space: {space}")
    index = nmslib.init(method=method, space=space)
    all_images = []

    qty = 0
    for p in tqdm(files):
        data = pickle.load(open(p, "rb"))
        images = data["images"]
        embeddings = data["embeddings"]
        assert len(images) == len(embeddings)
        #tqdm.write(f"embeddings shape: {embeddings.shape}")
        print('shape', embeddings.shape)

        ids = np.arange(qty, qty + len(embeddings))
        qty += len(embeddings)
        index.addDataPointBatch(embeddings, ids)
        all_images += images

        del embeddings, images, data

    # Create index
    index.createIndex(print_progress=True)

    return index, all_images
```

**Comment by xiankgx** [comment]
Ok noted, thank you so much for the quick response. Glad to be of help.

**Comment by danwhale** [comment]
This is really actual bug, and there is no simple warning in readme/manual to handle this properly.

As far I don't find this issue, with explicitly definition of IDs for adding batch, I can't solve my task and thought that I have wrong hyperparameters for my index. 

As more than year passed since this issue were created and such bug still exist may be simply include some warning in readme or python bindings API descritpion?

**Comment by searchivarius** [comment]
@danwhale sure, which place in API would be best?

**Comment by danwhale** [comment]
I think in docs it is necessary:

<img width="713" alt="image" src="https://user-images.githubusercontent.com/23121997/187173460-67fc0328-f66f-4804-8c53-7c6d03dcf53c.png">

But to avoid any naive usage, one able to trow a Warning message on each call of such method.

**Comment by searchivarius** [comment]
Ok, soon, I am gonna work on this. Sorry, it have been extremely busy times.
